pkgs/collection/lib/src/iterable_extensions.dart

```diff
@@ -1056,3 +1150,206 @@ extension ComparatorExtension<T> on Comparator<T> {
         return result;
       };
 }
+
+/// Implementation of [IterableExtension.separated].
+///
+/// Optimizes direct accesses.
+class _SeparatedIterable<T> extends Iterable<T> {
+  final T _separator;
+  final Iterable<T> _elements;
+
+  static const int _afterFlag = 1 << 0;
+  static const int _beforeFlag = 1 << 1;
+
+  /// Two bit-flags, for whether the `before` and `after` arguments were `true`.
+  final int _flags;
+
+  _SeparatedIterable(this._elements, this._separator, bool before, bool after)
+      : _flags = (before ? _beforeFlag : 0) + (after ? _afterFlag : 0);
+
+  @override
+  bool get isEmpty => _elements.isEmpty;
+  @override
+  bool get isNotEmpty => _elements.isNotEmpty;
+  @override
+  int get length {
+    var length = _elements.length;
+    if (length != 0) {
+      length = length * 2 - 1 + (_flags & 1) + (_flags >> 1);
+    }
+    return length;
+  }
+
+  @override
+  T elementAt(int index) {
+    RangeError.checkNotNegative(index, 'index');
+    // Figure out which element must exist in [_elements] for this index
+    // to exist in the separated output.
+    var indexWithoutBefore = index - (_flags >> 1);
+    var elementIndex = indexWithoutBefore ~/ 2; // Rounds both -1 and 1 to 0.
+    if (indexWithoutBefore.isEven) {
+      // It's an element.
+      return _elements.elementAt(elementIndex);
+    }
+    // It's a separator after that element (or before the first element).
+    // Check if that element exists, unless the `_afterFlag` is set,
+    // in which case to check if the next element exists by adding 1
+    // to elementIndex.
+    // (But if `index` is zero, it's the before separator, so it should
+    // check that a first element exists.)
+    if (index != 0) {
+      assert(_afterFlag == 1);
+      elementIndex += (_flags ^ _afterFlag) & _afterFlag;
+    }
+    _elements.elementAt(elementIndex); // If throws, not an element.
+    return _separator;
+  }
+
+  @override
+  T get first {
+    if (_flags & _beforeFlag == 0) return _elements.first;
+    if (_elements.isNotEmpty) return _separator;
+    throw StateError('No element');
+  }
+
+  @override
+  T get last {
+    if (_flags & _afterFlag == 0) return _elements.last;
+    if (_elements.isNotEmpty) return _separator;
+    throw StateError('No element');
+  }
+
+  @override
+  Iterable<T> take(int count) {
+    if (count == 0) return Iterable<T>.empty();
+    var beforeCount = _flags >> 1;
+    if (count == 1) {
+      if (beforeCount == 0) {
+        return _elements.take(1);
+      }
+      // return Iterable<T>.value(_separator); // Why you no exist?!
+      return DelegatingIterable<T>([_separator]);
```

Comment:
Are we wrapping this because we're worried about a caller detecting it's a list and starting to rely on that detail?